pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,398 @@
+	if result.Canceled {
+		c.logger.Error("failed to put chunk, task is canceled", zap.String("error", result.Error))
+		return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+	}
+
+	if result.HandledChunkID != expectedChunkID {
+		c.logger.Info("remote worker may restart, retry to put chunk",
+			zap.Uint64("expected chunkID", expectedChunkID),
+			zap.Uint64("handled chunkID", result.HandledChunkID),
+			zap.Uint64("flushed chunkID", result.FlushedChunkID))
+
+		nextChunkID := result.HandledChunkID + 1
+		for nextChunkID <= expectedChunkID {
+			url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, nextChunkID)
+
+			buf, err := c.chunksCache.get(nextChunkID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.logger.Info("retry to put chunk", zap.Uint64("chunkID", nextChunkID))
+
+			data, err := sendRequest(ctx, c.httpClient, "PUT", url, buf)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			result = new(PutChunkResult)
+			err = json.Unmarshal(data, result)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if result.Canceled {
+				c.logger.Error("failed to put chunk, task is canceled", zap.String("error", result.Error))
+				return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+			}
+
+			nextChunkID = result.HandledChunkID + 1
+		}
+	}
+
+	state := c.state.Load()
+	// We can clean the cache of chunks that the remote worker has flushed.
+	lastFlushedChunkID := state.FlushedChunkID + 1
+	for lastFlushedChunkID <= result.FlushedChunkID {
+		err := c.chunksCache.clean(lastFlushedChunkID)
+		if err != nil {
+			c.logger.Info("failed to clean chunk cache",
+				zap.Uint64("chunkID", lastFlushedChunkID),
+				zap.Error(err))
+		}
+		lastFlushedChunkID++
+	}
+
+	c.state.Store(&chunksState{
+		FlushedChunkID: result.FlushedChunkID,
+		HandledChunkID: result.HandledChunkID,
+	})
+	return nil
+}
+
+// sendFlushToRemote send flush requst to remote worker.
+//
+// The remote worker should return `FlushedChunkID`, which should be equal to `LastChunkID`.
+// If not, it means that the remote worker has restarted and we should resend the chunk from
+// `FlushedChunkID` to the `LastChunkID`.
+func (c *chunkSender) sendFlushToRemote(ctx context.Context) error {
+	for {
+		url := fmt.Sprintf(flushURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id)
+
+		data, err := sendRequest(ctx, c.httpClient, "POST", url, nil)
+		if err != nil {
+			c.logger.Error("failed to flush", zap.Error(err))
+			return errors.Trace(err)
+		}
+
+		result := new(FlushResult)
+		err = json.Unmarshal(data, result)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if result.Canceled || result.Error != "" {
+			c.logger.Error("failed to flush",
+				zap.Bool("canceled", result.Canceled),
+				zap.String("error", result.Error))
+			return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+		}
+
+		if result.Finished {
```

Comment:
I wonder how does remote server know a task is finished? I think only the client side knows how many task chunks will be generated.

Reply:
The `Finished` means all the data has been ingested to tikv, the `Finished` here should not be true. I add comment for field in `PutChunkResult` and `FlushResult`.